### DISCUSSION TITLE
Unable to receive signal from org.freedesktop.DBus

### DISCUSSION THREAD

**Discussion by aacebedo** [diwic/dbus-rs#423] [Q&A] [answered]
Hi,

I am trying to receive a signal "NameAcquired" from the session bus with tokio. My goal is to receive a signal when a specific object is added to the bus. I took the example and modified it as follow:
```
use dbus_tokio::connection;
use dbus::nonblock;
use std::time::Duration;
use dbus::message::MatchRule;

#[tokio::main]
pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (resource, conn) = connection::new_session_sync()?;

    let _handle = tokio::spawn(async {
        let err = resource.await;
        panic!("Lost connection to D-Bus: {}", err);
    });

    let mr = MatchRule::new_signal("org.freedesktop.DBus", "NameAcquired");
    let incoming_signal = conn.add_match(mr).await?.cb(|_, (source,): (String,)| {
        println!("Hello from {} happened on the bus!", source);
        true
    });

    loop {}

    conn.remove_match(incoming_signal.token()).await?;

    unreachable!()
}
```

However it does not work. I can see the signals with dbus-monitor (for instance by launching vlc) but my program does not react. 

Any idea of what I am missing here?

**Comment by diwic** [comment, marked as answer]
Unless tokio has done something weird with its main function, `loop {}` busy-loops with 100% CPU without giving any chance for anything else to run, including signal processing. You need something like `future::pending::<()>().await` instead.

**Reply by aacebedo** [reply]
Actually I thought about the infinite loop and tried the following with the same issue:

```
use dbus_tokio::connection;
use dbus::nonblock;
use std::time::Duration;
use dbus::message::MatchRule;

#[tokio::main]
pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (resource, conn) = connection::new_session_sync()?;

    let _handle = tokio::spawn(async {
        let err = resource.await;
        panic!("Lost connection to D-Bus: {}", err);
    });

    let mr = MatchRule::new_signal("org.freedesktop.DBus", "NameAcquired");
    let incoming_signal = conn.add_match(mr).await?.cb(|_, (source,): (String,)| {
        println!("Hello from {} happened on the bus!", source);
        true
    });

        tokio::join!(tokio::spawn(async {
        loop {
            println!("sleep");
            tokio::time::sleep(Duration::from_millis(1000)).await;
        }
    }));


    conn.remove_match(incoming_signal.token()).await?;

    unreachable!()
}
```

In the dbus-monitor I see this:

```
signal time=1675099416.729247 sender=org.freedesktop.DBus -> destination=:1.733 serial=3 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string "org.mpris.MediaPlayer2.vlc"

```

**Reply by aacebedo** [reply]
How does the cb given to the result of the add_match is processed by tokio? Through the _handle task?

**Reply by diwic** [reply]
I've tried to debug your issue and I think it's something with the `NameAcquired` signal only being sent to the application that acquires the name, not to other applications.

And that dbus-monitor have special calls (BecomeMonitor, eavesdrop etc) to get these signals also. But I'm not sure.

**Reply by diwic** [reply]
This works better because `NameOwnerChanged` is sent to the entire bus, not to a specific application:

    let mr = MatchRule::new_signal("org.freedesktop.DBus", "NameOwnerChanged");
    let incoming_signal = conn.add_match(mr).await?.cb(|_, (name, oldowner, newowner): (String, String, String)| {
        println!("Changed owner: {} from {} to {}", name, oldowner, newowner);
        true
    });

**Reply by aacebedo** [reply]
That's working ! Thank you for the time you took to troubleshoot that. I didn't know that a signal could be sent to a specific recipient.
